consensus/transaction_solidifier/transaction_solidifier.c

```diff
@@ -0,0 +1,373 @@
+
+static retcode_t propogate_solid_transactions(
+    transaction_solidifier_t *const ts);
+
+static retcode_t check_solidity_do_func(tangle_t *const tangle,
+                                        flex_trit_t *hash,
+                                        iota_stor_pack_t *pack, void *data,
+                                        bool *should_branch, bool *should_stop);
+
+/*
+ * Auxiliary structs
+ */
+
+typedef struct check_solidity_do_func_params_s {
+  transaction_solidifier_t *ts;
+  bool is_milestone;
+} check_solidity_do_func_params_t;
+
+static void *spawn_solid_transactions_propogation(void *arg) {
+  transaction_solidifier_t *ts = (transaction_solidifier_t *)arg;
+  while (ts->running) {
+    if (propogate_solid_transactions(ts) != RC_OK) {
+      log_critical(TRANSACTION_SOLIDIFIER_LOGGER_ID,
+                   "Failed in solid transaction propagation %s\n");
+    }
+    sleep(SOLID_PROPAGATION_INTERVAL);
+  }
+  return NULL;
+}
+
+static retcode_t propogate_solid_transactions(
+    transaction_solidifier_t *const ts) {
+  retcode_t ret;
+  hash_set_t transactions_to_propagate;
+
+  lock_handle_lock(&ts->lock);
+  if ((ret = hash_set_for_each(&ts->newly_set_solid_transactions, hash_set_add,
+                               (void *)&transactions_to_propagate))) {
+    return ret;
+  }
+
+  lock_handle_unlock(&ts->lock);
+
+  iota_stor_pack_t hash_pack;
+  if ((ret = hash_pack_init(&hash_pack, 10)) != RC_OK) {
+    return ret;
+  }
+
+  hash_set_entry_t *curr_entry = NULL;
+  hash_set_entry_t *tmp_entry = NULL;
+  flex_trit_t *curr_approver_hash;
+  uint16_t approver_index;
+  bool is_approver_solid;
+  HASH_ITER(hh, transactions_to_propagate, curr_entry, tmp_entry) {
+    if (!ts->running) {
+      break;
+    }
+
+    hash_pack_reset(&hash_pack);
+    if ((ret = iota_tangle_transaction_load_hashes_of_approvers(
+             ts->tangle, curr_entry->hash, &hash_pack))) {
+      return ret;
+    }
+    for (approver_index = 0; approver_index < hash_pack.num_loaded;
+         ++approver_index) {
+      curr_approver_hash = (flex_trit_t *)hash_pack.models[approver_index];
+      if ((ret = check_transaction_and_update_solid_state(
+               ts, curr_approver_hash, &is_approver_solid)) != RC_OK) {
+        return ret;
+      }
+      if (is_approver_solid) {
+        if ((ret = iota_tangle_transaction_update_solid_state(
+                 ts->tangle, curr_approver_hash, true)) != RC_OK) {
+          return ret;
+        }
+
+        if ((ret = add_new_solid_transaction(ts, curr_approver_hash)) !=
+            RC_OK) {
+          return ret;
+        }
+      }
+    }
+  }
+  return RC_OK;
+}
+
+void iota_consensus_transaction_solidifier_init(
+    transaction_solidifier_t *const ts, tangle_t *const tangle,
+    requester_state_t *const requester) {
+  ts->tangle = tangle;
+  ts->requester = requester;
+  ts->running = false;
+  ts->newly_set_solid_transactions = NULL;
```

Comment:
you never set an actual value here? or do the HASH macros take care of this?

Reply:
They do